File: python/32_tebd.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
(Simplified) TEBD simulation of random quantum circuits
"""

import numpy as np


def main():
    n = 16  # number of qubits
    depth = 16  # number of layers
    max_dim = 4  # for MPS with truncation
    cutoff = 1e-10  # for MPS with truncation

    # two-qubit gate: CNOT
    cnot = np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0]]).reshape(
        (2, 2, 2, 2)
    )

    # state vector |0...0>
    state = np.zeros((2**n, 1))
    state[0] = 1
    state = state.reshape((2,) * n)

    # MPS |0...0> without truncation
    mps0 = []
    for i in range(n):
        mps0.append(np.array([[[1], [0]]]))
    print(
        f"mps0/1 initial virtual bond dimensions: {[mps0[i].shape[2] for i in range(n - 1)]}"
    )

    # MPS |0...0> with truncation
    mps1 = mps0.copy()

    for k in range(depth):
        # random single-qubit rotations
        for pos in range(n):
            alpha = np.random.uniform(0, 2 * np.pi)
            theta = np.random.uniform(0, np.pi)
            phi = np.random.uniform(0, 2 * np.pi)
            U = np.exp(1j * alpha) * np.array(
                [
                    [np.cos(theta / 2), -np.exp(1j * phi) * np.sin(theta / 2)],
                    [np.exp(-1j * phi) * np.sin(theta / 2), np.cos(theta / 2)],
                ]
            )

            ss_op = list(range(2))
            ss_in = list(range(2, 2 + n))
            ss_to = list(range(2, 2 + n))
            ss_in[pos] = ss_op[1]
            ss_to[pos] = ss_op[0]
            state = np.einsum(U, ss_op, state, ss_in, ss_to)

            mps0[pos] = np.einsum("ij,kjm->kim", U, mps0[pos])
            mps1[pos] = np.einsum("ij,kjm->kim", U, mps1[pos])

        # cnot gates
        for i in range(k // 2, n - 1, 2):
            pos = [i, i + 1]

            ss_op = list(range(4))
            ss_in = list(range(4, 4 + n))
            ss_to = list(range(4, 4 + n))
            ss_in[pos[0]] = ss_op[2]
            ss_in[pos[1]] = ss_op[3]
            ss_to[pos[0]] = ss_op[0]
            ss_to[pos[1]] = ss_op[1]
            state = np.einsum(cnot, ss_op, state, ss_in, ss_to)

            t = np.einsum("ijkl,mkn->mijln", cnot, mps0[pos[0]])
            t = np.einsum("ijklm,mln->ijkn", t, mps0[pos[1]])
            t = t.reshape(t.shape[0] * t.shape[1], t.shape[2] * t.shape[3])
            U, S, Vt = np.linalg.svd(t, full_matrices=False)
            S = S[S > cutoff]
            S_sqrt = np.diag(np.sqrt(S))
            mps0[pos[0]] = (U[:, : len(S)] @ S_sqrt).reshape(-1, 2, len(S))
            mps0[pos[1]] = (S_sqrt @ Vt[: len(S), :]).reshape(len(S), 2, -1)

            t = np.einsum("ijkl,mkn->mijln", cnot, mps1[pos[0]])
            t = np.einsum("ijklm,mln->ijkn", t, mps1[pos[1]])
            t = t.reshape(t.shape[0] * t.shape[1], t.shape[2] * t.shape[3])
            U, S, Vt = np.linalg.svd(t, full_matrices=False)
            S = S[S > cutoff]
            S = S[: min(len(S), max_dim)]
            S_sqrt = np.diag(np.sqrt(S))
            mps1[pos[0]] = (U[:, : len(S)] @ S_sqrt).reshape(-1, 2, len(S))
            mps1[pos[1]] = (S_sqrt @ Vt[: len(S), :]).reshape(len(S), 2, -1)

        state0 = mps0[0].reshape(2, -1)
        for i in range(1, n):
            state0 = np.einsum("ij,jkl->ikl", state0, mps0[i])
            state0 = state0.reshape(state0.shape[0] * state0.shape[1], state0.shape[2])
        state0 = state0.reshape(-1)

        state1 = mps1[0].reshape(2, -1)
        for i in range(1, n):
            state1 = np.einsum("ij,jkl->ikl", state1, mps1[i])
            state1 = state1.reshape(state1.shape[0] * state1.shape[1], state1.shape[2])
        state1 = state1.reshape(-1)

        print(
            f"step: {k} fidelity: {np.abs(np.vdot(state, state0)) ** 2}, {np.abs(np.vdot(state, state1)) ** 2}"
        )

    print(
        f"mps0 final virtual bond dimensions: {[mps0[i].shape[2] for i in range(n - 1)]}"
    )
    print(
        f"mps1 final virtual bond dimensions: {[mps1[i].shape[2] for i in range(n - 1)]}"
    )


if __name__ == "__main__":
    main()

File: rust/src/bin/32_tebd.rs
use anyhow::Result;
use ndarray::{Array1, Array2, Array3, Array4, Ix3, Ix4, Ix5, array, s};
use ndarray_einsum::einsum;
use num_complex::Complex64 as C64;
use rand::Rng;
use std::f64::consts::PI;
use tnqc_springschool::{EasySVD, MapStrToAnyhowErr};

fn iexp(theta: f64) -> C64 {
    C64::new(theta.cos(), theta.sin())
}

// (Simplified) TEBD simulation of random quantum circuits

fn main() -> Result<()> {
    let n: usize = 16;
    let depth: usize = 16;
    let max_dim: usize = 4;
    let cutoff: f64 = 1e-10;

    // two-qubit gate: CNOT
    let cnot: Array4<C64> = array![
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 1.0, 0.0, 0.0],
        [0.0, 0.0, 0.0, 1.0],
        [0.0, 0.0, 1.0, 0.0]
    ]
    .mapv(|x: f64| C64::from(x))
    .into_shape_with_order((2, 2, 2, 2))?;

    // state vector |0...0>
    let mut state = Array1::<C64>::zeros([1usize << n]);
    state[0] = 1.0.into();

    // MPS |0...0> without truncation
    let mut mps0: Vec<Array3<C64>> = (0..n)
        .map(|_| array![[[1.0.into()], [0.0.into()]]])
        .collect();
    print!("mps0/1 initial virtual bond dimensions: [",);
    for i in 0..(n - 1) {
        print!("{}, ", mps0[i].dim().2);
    }
    println!("]\n");

    // MPS |0...0> with truncation
    let mut mps1 = mps0.clone();

    let mut rng = rand::rng();

    for k in 0..depth {
        // random single-qubit rotations
        for pos in 0..n {
            let alpha = rng.random_range(0.0..2.0 * PI);
            let theta = rng.random_range(0.0..PI);
            let phi = rng.random_range(0.0..2.0 * PI);

            let u = iexp(alpha)
                * array![
                    [
                        C64::from((theta / 2.0).cos()),
                        -iexp(phi) * C64::from((theta / 2.0).sin())
                    ],
                    [
                        iexp(-phi) * C64::from((theta / 2.0).sin()),
                        C64::from((theta / 2.0).cos())
                    ]
                ];

            {
                // due to ndarray_einsum limitation, we apply gate by coonverting vector into fixed shape tensor
                let state_3d =
                    state.into_shape_with_order((1usize << pos, 2, 1usize << (n - pos - 1)))?;
                let state_3d = einsum("ij,kjm->kim", &[&u, &state_3d])
                    .map_str_err()?
                    .into_dimensionality::<Ix3>()?;
                state = state_3d.flatten().into_owned();
            }
            {
                mps0[pos] = einsum("ij,kjm->kim", &[&u, &mps0[pos]])
                    .map_str_err()?
                    .into_dimensionality::<Ix3>()?;
            }
            {
                mps1[pos] = einsum("ij,kjm->kim", &[&u, &mps1[pos]])
                    .map_str_err()?
                    .into_dimensionality::<Ix3>()?;
            }
        }

        // cnot gates
        for i in ((k / 2)..(n - 1)).step_by(2) {
            {
                // due to ndarray_einsum limitation, we apply gate by coonverting vector into fixed shape tensor
                let state_4d = state.to_shape((1usize << i, 2, 2, 1usize << (n - i - 2)))?;
                let state_4d = einsum("ijkl,mkln->mijn", &[&cnot, &state_4d])
                    .map_str_err()?
                    .into_dimensionality::<Ix4>()?;
                state = state_4d.flatten().into_owned();
            }
            {
                let sub0 = einsum("ijkl,mkn->mijln", &[&cnot, &mps0[i]])
                    .map_str_err()?
                    .into_dimensionality::<Ix5>()?;
                let sub0 = einsum("ijklm,mln->ijkn", &[&sub0, &mps0[i + 1]])
                    .map_str_err()?
                    .into_dimensionality::<Ix4>()?;
                let shape = sub0.dim();
                let tmat = sub0.to_shape((shape.0 * shape.1, shape.2 * shape.3))?;
                let (u, s, vt) = tmat.thin_svd()?;
                let rank_new = s
                    .iter()
                    .position(|&x| x <= cutoff * s[0])
                    .unwrap_or(s.len());
                let s = s.slice(s![0..rank_new]);
                let smat = Array2::from_diag(&s.mapv(|x| C64::from(x.sqrt())));
                mps0[i] = (u.slice(s![.., 0..rank_new]))
                    .dot(&smat)
                    .into_shape_clone((shape.0, shape.1, rank_new))?;
                mps0[i + 1] = smat
                    .dot(&(vt.slice(s![0..rank_new, ..])))
                    .into_shape_clone((rank_new, shape.2, shape.3))?;
            }
            {
                let sub1 = einsum("ijkl,mkn->mijln", &[&cnot, &mps1[i]])
                    .map_str_err()?
                    .into_dimensionality::<Ix5>()?;
                let sub1 = einsum("ijklm,mln->ijkn", &[&sub1, &mps1[i + 1]])
                    .map_str_err()?
                    .into_dimensionality::<Ix4>()?;
                let shape = sub1.dim();
                let tmat = sub1.to_shape((shape.0 * shape.1, shape.2 * shape.3))?;
                let (u, s, vt) = tmat.thin_svd()?;
                let rank_new = s
                    .iter()
                    .position(|&x| x <= cutoff * s[0])
                    .unwrap_or(s.len())
                    .min(max_dim);
                let s = s.slice(s![0..rank_new]);
                let smat = Array2::from_diag(&s.mapv(|x| C64::from(x.sqrt())));
                mps1[i] = (u.slice(s![.., 0..rank_new]))
                    .dot(&smat)
                    .into_shape_clone((shape.0, shape.1, rank_new))?;
                mps1[i + 1] = smat
                    .dot(&(vt.slice(s![0..rank_new, ..])))
                    .into_shape_clone((rank_new, shape.2, shape.3))?;
            }
        }

        let shape = mps0[0].dim();
        let mut state0 = mps0[0].to_shape((shape.0 * shape.1, shape.2))?.to_owned();
        for i in 1..n {
            let state = einsum("ij,jkl->ikl", &[&state0, &mps0[i]])
                .map_str_err()?
                .into_dimensionality::<Ix3>()?;
            let shape = state.dim();
            state0 = state.into_shape_clone((shape.0 * shape.1, shape.2))?;
        }
        let state0 = state0.flatten();

        let shape = mps1[0].dim();
        let mut state1 = mps1[0].to_shape((shape.0 * shape.1, shape.2))?.to_owned();
        for i in 1..n {
            let state = einsum("ij,jkl->ikl", &[&state1, &mps1[i]])
                .map_str_err()?
                .into_dimensionality::<Ix3>()?;
            let shape = state.dim();
            state1 = state.into_shape_clone((shape.0 * shape.1, shape.2))?;
        }
        let state1 = state1.flatten();

        let state_conj = state0.mapv(|x| x.conj());

        println!(
            "step: {} fidelity: {}, {}",
            k,
            state_conj.dot(&state0).norm_sqr(),
            state_conj.dot(&state1).norm_sqr(),
        );
    }

    print!("mps0 final virtual bond dimensions: [",);
    for i in 0..(n - 1) {
        print!("{}, ", mps0[i].shape()[2]);
    }
    println!("]\n");
    print!("mps1 final virtual bond dimensions: [",);
    for i in 0..(n - 1) {
        print!("{}, ", mps1[i].shape()[2]);
    }
    println!("]\n");
    Ok(())
}
